scripts/gen_dockerfile.py

```diff
@@ -0,0 +1,250 @@
+
+"""Generate a Dockerfile and helper files for a Python application."""
+
+import argparse
+import collections
+import functools
+import io
+import os
+import re
+import sys
+
+import yaml
+
+import validation_utils
+
+
+# Validate characters for dockerfile image names.
+#
+# This roots out obvious mistakes, the full gory details are here:
+# https://github.com/docker/distribution/blob/master/reference/regexp.go
+IMAGE_REGEX = re.compile(r"""(?x)
+    ^
+    [a-zA-Z0-9]          # First char must be alphanumeric
+    [a-zA-Z0-9-_./:@+]*  # Punctuation allowed after that
+    $
+""")
+
+# `entrypoint` is specified as free-form text parsed as a unix shell
+# command line, which limits the sanity checking possible.  We
+# disallow newlines and control characters which would break the
+# Dockerfile format.
+PRINTABLE_REGEX = re.compile(r"""^[^\x00-\x1f]*$""")
+
+# Map from app.yaml "python_version" to {python_version} in Dockerfile
+PYTHON_INTERPRETER_VERSION_MAP = {
+    '': '',  # == 2.7
+    '2': '',  # == 2.7
+    '3': '3.5',
+    '3.4': '3.4',
+    '3.5': '3.5',
+    '3.6': '3.6',
+}
+
+# File templates.
+# Designed to exactly match the current output of 'gcloud app gen-config'
+DOCKERFILE_TEMPLATE = """\
+FROM {base_image}
+LABEL python_version=python{dockerfile_python_version}
+RUN virtualenv --no-download /env -p python{dockerfile_python_version}
+
+# Set virtualenv environment variables. This is equivalent to running
+# source /env/bin/activate
+ENV VIRTUAL_ENV /env
+ENV PATH /env/bin:$PATH
+{optional_requirements_txt}ADD . /app/
+{optional_entrypoint}"""
+
+DOCKERFILE_REQUIREMENTS_TXT = """\
+ADD requirements.txt /app/
+RUN pip install -r requirements.txt
+"""
+
+DOCKERFILE_ENTRYPOINT_TEMPLATE = """\
+CMD {entrypoint}
+"""
+
+DOCKERIGNORE = """\
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+.dockerignore
+Dockerfile
+.git
+.hg
+.svn
+"""
+
+# Validated application configuration
+AppConfig = collections.namedtuple(
+    'AppConfig',
+    'base_image dockerfile_python_version entrypoint has_requirements_txt'
+)
+
+
+def get_app_config(raw_config, args):
+    """Read and validate the application runtime configuration.
+
+    Args:
+        raw_config (dict): deserialized app.yaml
+        args (argparse.Namespace): command line flags
+
+    Returns:
+        AppConfig: valid configuration
+    """
+    # Examine app.yaml
+    if not isinstance(raw_config, dict):
```

Comment:
That's fine - "I'm dealing with user input I can't trust" is a valid answer, I just wanted to understand the motivation. Now, my key issue here is asserting `dict` when any valid mapping will do - could you "relax" this to `isinstance(raw_config, types.Mapping)`?